huntsman/observatory.py

```diff
@@ -472,43 +355,217 @@ def _create_autoguider(self):
+
+                # Check the next exposure time is within limits
+                if next_exptime > max_exptime:
+                    self.logger.debug(f'Suggested exposure time for {cam_name}'
+                                      f' is too long: {next_exptime}.')
+                    if not self.past_midnight:
+                        finished[cam_name] = True  # Its getting darker, so finish
+                        continue
+                    next_exptime = max_exptime
+                elif next_exptime < min_exptime:
+                    self.logger.debug(f'Suggested exposure time for {cam_name}'
+                                      f' is too short: {next_exptime}.')
+                    if self.past_midnight:
+                        finished[cam_name] = True  # Its getting lighter, so finish
+                        continue
+                    next_exptime = min_exptime
+
+                # Update the next exposure time
+                exptimes[cam_name].append(next_exptime)
+
+            # Check if all the exposures in this loop are too bright
+            if self.past_midnight:
+                if all_too_faint:
+                    self.logger.debug('All exposures are too faint. Waiting 30 seconds...')
+                    time.sleep(30)
+            else:
+                if all_too_bright:
+                    self.logger.debug('All exposures are too bright. Waiting 30 seconds...')
+                    time.sleep(30)
+
+        # Return the exposure times
+        return exptimes
+
+    def _autoflat_mean_counts(self, filename, bias, min_counts=1):
+        """ Read the data and calculate a clipped-mean count rate.
 
-        self.logger.debug("Flat-field observation: {}".format(flat_obs))
+        Args:
+            filename (str): The filename containing the data.
+            bias (float): The bias level to subtract from the image.
+            min_counts (float): The minimum count rate returned by this funtion.
+        """
+        try:
+            data = fits.getdata(filename)
+        except FileNotFoundError:
+            data = fits.getdata(filename + '.fz')
+
+        # Calculate average counts per pixel
+        mean_counts, _, _ = stats.sigma_clipped_stats(data-bias)
+        mean_counts = max(min_counts, mean_counts - bias)
+        return mean_counts
+
+    def _autoflat_next_exptime(self, previous_exptime, elapsed_time, target_counts, mean_counts):
+        """Calculate the next exposure time for the flat fields, accounting
+        for changes in sky brightness."""
+        exptime = previous_exptime * (target_counts / mean_counts)
+        sky_factor = 2.0 ** (elapsed_time / 180.0)
+        if self.past_midnight:
+            exptime = exptime / sky_factor
+        else:
+            exptime = exptime * sky_factor
+        return round(exptime.to_value(u.second)) * u.second
+
+    def _take_flat_observation(self, exptimes, observation, fits_headers=None, dark=False):
```

Comment:
Need to add `safety_func`.
```suggestion
    def _take_flat_observation(self, exptimes, observation, safety_func, fits_headers=None, dark=False):
```